pandas/core/indexing.py

```diff
@@ -906,38 +922,34 @@ def _multi_take(self, tup):
         """
         try:
             o = self.obj
-            d = {a: self._convert_for_reindex(t, axis=o._get_axis_number(a))
-                 for t, a in zip(tup, o._AXIS_ORDERS)}
-            return o.reindex(**d)
-        except(KeyError, IndexingError):
-            raise self._exception
+            d = {}
+            for key, axis in zip(tup, o._AXIS_ORDERS):
+                ax = o._get_axis(axis)
+                # Have the index compute an indexer or return None
+                # if it cannot handle:
+                indexer, keyarr = ax._convert_listlike_indexer(key,
+                                                               kind=self.name)
+                # We only act on all found values:
+                if indexer is not None and (indexer != -1).all():
+                    self._validate_read_indexer(key, indexer, axis)
+                    d[axis] = (ax[indexer], indexer)
+                    continue
+
+                # If we are trying to get actual keys from empty Series, we
+                # patiently wait for a KeyError later on - otherwise, convert
+                if len(ax) or not len(key):
+                    key = self._convert_for_reindex(key, axis)
+                indexer = ax.get_indexer_for(key)
+                keyarr = ax.reindex(keyarr)[0]
+                self._validate_read_indexer(keyarr, indexer,
+                                            o._get_axis_number(axis))
+                d[axis] = (keyarr, indexer)
+            return o._reindex_with_indexers(d, copy=True, allow_dups=True)
+        except(KeyError, IndexingError) as detail:
```

Comment:
not sure if its worth it, but is it possible to surround only the op that could raise the IndexingError? (with the try/except)

Reply:
Didn't address this. I agree that'd be nice, but at a glance it looks like there are 3-4 places that could raise one of those errors. Seems tricky